### PR TITLE
Arrays - rearrange array

### PR DESCRIPTION
**Pull request by DevratAgarwal**
## Checklist
- Solution is tested on sphere engine <br/>
    `Note:` Please copy only the solution part. Do not copy/paste the template code. That way, any issues with template code can be caught early on.
    - [ ] Author
    - [x] Reviewer
- PR title is correct - [PR title conventions](https://github.com/DSA-solutions-the10xacademy/JavaScript-DSA-Solutions#title-for-a-pull-request)
    - [ ] Author
    - [x] Reviewer
- Folder/File naming is correct - [Naming conventions](https://github.com/DSA-solutions-the10xacademy/JavaScript-DSA-Solutions#naming-of-files-and-folders)
    - [ ] Author
    - [ ] Reviewer
- Coding conventions are followed - [Coding conventions](https://github.com/DSA-solutions-the10xacademy/JavaScript-DSA-Solutions#coding-conventions)
    - [ ] Author
    - [ ] Reviewer
- Test cases (action item for **reviewer**) <br/>
    - To be ensured by checking the problem on sphere-engine
    - If any test cases are missing, please add them - [how to add new test cases](https://github.com/DSA-solutions-the10xacademy/JavaScript-DSA-Solutions#adding-new-test-cases)
    - [x] There are at least 2 test cases
    - [x] There is at least 1 test case that is different from the sample test cases

### CODE REVIEW

**Comment by rpapavan97** [comment]
@DevratAgarwal , add ".js" file extension to solution file, and also give spaces around binary operators (like "=" while defining variables)

**Comment by DevratAgarwal** [comment]
@rpapavan97 done

**Review by rpapavan97 — APPROVED** [review]

**Comment by vinthaivendhan10x** [comment]
@sailesh-ramayanam kindly Review

File: solutions/arrays/rearrange-array.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+for (let idx = 0; idx < arraySize; idx++) {
+  nums.push(parseInt(readLine()));
+}
+let index = [];
+for (let idx = 0; idx < arraySize; idx++) {
+  index.push(parseInt(readLine()));
+}
+let answer = [];
+for (let idx = 0; idx < arraySize; idx++) {
+  answer.splice(index[idx], 0, nums[idx]);
```
Comment (sailesh-ramayanam) [inline solutions/arrays/rearrange-array.js:24]: @DevratAgarwal Please don't use `splice`. This is very specific to `Javascript`. Students will assume that this is an `O(1)` operation. Please do this without `splice` operation.

**Review by sailesh-ramayanam — CHANGES_REQUESTED** [review]

File: solutions/arrays/rearrange-array.js (hunk 1 above)
Comment (DevratAgarwal) [reply solutions/arrays/rearrange-array.js]: @sailesh-ramayanam done

File: solutions/arrays/rearrange-array.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+let target = [];
+for (let idx = 0; idx < arraySize; idx++) {
+    let currElement = nums[idx];
+    let targetIndex = indices[idx];
+    // We need to insert currElement at targetIndex
+    // But, we cannot create additional space in the middle of an array
+    // So, we will first insert currElement at the end of target
+    // Then, we will shift it towards left, one index at a time
```
Comment (sailesh-ramayanam) [inline solutions/arrays/rearrange-array.js:32]: @DevratAgarwal I changed the algorithm. I feel the new one is simpler and easier to understand. Also, this uses a constant auxiliary space (primary reason for change is that this is easier to understand).

**Review by sailesh-ramayanam — APPROVED** [review]

**Pull request merged**